madmom/features/beats.py

```diff
@@ -1238,6 +1238,380 @@ def add_arguments(parser, min_bpm=MIN_BPM, max_bpm=MAX_BPM,
+        histogram = self.tempo_estimator.interval_histogram(induction_window)
+        # get N most likely tempi
+        from .tempo import detect_tempo
+        tempi = detect_tempo(histogram, self.fps)[:self.TEMPI, 0]
+        # convert tempi to intervals
+        intervals = 60.0 * self.fps / tempi
+        # induct agents for each interval
+        for interval in intervals.astype(int):
+            self.induct_agents(activations[:interval], interval)
+        # iterate through all activations
+        for idx, activation in enumerate(activations):
+            # process activation for each agent
+            new_agents = []
+            for agent in self.agents:
+                # calculate number of frames to look around the prediction
+                # TODO: Should this be calculated as a @property in agent?
+                outer_frames = int(agent.interval * agent.OUTER_WINDOW)
+                # skip if not the time yet for this agent
+                if agent.prediction + outer_frames != idx:
+                    continue
+                # get activations of outer windows surrounding the prediction
+                context = activations[max(0, idx - outer_frames * 2):idx]
+                # process agent and extend new agents
+                new_agents.extend(agent.process(context, idx))
+            # TODO: Those lines are the same for offline/online: refactor?
+            # append new agents to agents list
+            self.agents.extend(new_agents)
+            # sort all agents by score
+            self.agents.sort(key=lambda a: a.score, reverse=True)
+            # remove duplicates by using the agents __eq__ method
+            self.agents = list(dict.fromkeys(self.agents))
+            # kill worst agents if too many
+            self.agents = self.agents[:self.MAX_AGENTS]
+        # return beats of best agent
+        return np.array(self.agents[0].beats) / self.fps
+
+    def process_online(self, activations, reset=True, **kwargs):
+        """
+        Detect the beats in the given activation function for online mode.
+
+        Parameters
+        ----------
+        activations : numpy array
+            Beat activation function.
+        reset : bool, optional
+            Reset the BeatTrackingProcessor to its initial state before
+            processing.
+
+        Returns
+        -------
+        beats : numpy array
+            Detected beat positions [seconds].
+
+        """
+        # reset to initial state
+        if reset:
+            self.reset()
+        beats_ = []
+        for activation in activations:
+            # shift buffer and put new activation at end of buffer
+            buffer = self.buffer(activation)
+            # induct agents after induction time has passed
+            if self.counter == self.INDUCTION_TIME * self.fps:
+                # create histogram of induction window
+                histogram = self.tempo_estimator.interval_histogram(buffer)
```

Comment:
I am not sure, but shouldn't `activation` be passed to the tempo estimator and not the buffer? In the latter case you compute the histogram for more than the last activation which is probably not what you want.

Reply:
Ah yeah, thats a good point and actually a larger topic ;)

Well, the concept for both offline and online is the same: Always use the first N seconds to estimate the tempo. Then introduce the agents (for offline at the beginning of the song, for online after those N seconds). From that point in time there is no more tempo estimation going on and the agents are themselves responsible to adapt to the changing speed of the music. That's why in the code the online tempo estimator is not used at all, only the offline tempo estimator is used once after those N seconds (Btw. thats also done in a similar way in all papers I found). That also means that for online the first N seconds don't have agents and therefore cannot produce beat predictions which probably leads to a worse result. For that reason in some papers they just ignore those first N induction seconds when evaluating - but I think that would be unfair to all the other tracking algorithms which are capable of producing predictions from the beginning.

One other problem we are facing here is that the online networks tend to give worse predictions in (approx.) the first 5 seconds of a song. And that is exactly the section used to estimate the base-tempi for the agents. So I already thought about how agents could get their tempo updated during runtime but it's difficult because each agent has its own tempo already from the beginning (since we are using multiple tempo estimations) and updates it on the fly. 

Also connected to that is the handling of abrupt tempo changes. Some solutions I found use a kind of "overwatcher" module which is only looking for sudden changes and makes another induction when necessary. I don't know yet whether we will need something like this.

So yeah, right now for offline I get 90% and for online 80% F-Score for the Ballroom set (didn't play much with the parameters yet). If we get some more ideas I will try them out ;)

Reply:
Yes, tempo estimation at the very beginning is indeed a problem. Relying on those estimations is thus probably not a very good idea. How about simply starting a fixed number of agents at the beginning of the song and initialise them with tempi distributed uniformly on a logarithmic tempo scale? Soon afterwards the "bad" agents will be killed/die anyways, right?

For example for downbeat tracking I limit the number of modelled tempi to 60 in order to prevent an excessively bloated state space. 60 tempi is enough to cover the whole tempo range from 40..250bpm. Please note, that the downbeat tracker is not able to use any other tempi.

I guess the agents are able to adapt their tempo quite quickly, so maybe this is a good workaround/extension and would render the need for tempo estimation completely.

Reply:
Ah, yeah interesting idea! I will test that (maybe also for offline) ;)